File: server/src/routes/sugerencias.ts
import { Router } from 'express'
import { pool } from '../db.js'

export const sugerenciasRouter = Router()

// Cola de Validación IA: sugerencias pendientes de revisión por un analista.
sugerenciasRouter.get('/', async (_req, res) => {
  const { rows } = await pool.query(`
    SELECT s.*, f.url, f.dominio, ti.nombre AS tipo_infraccion_nombre, ti.color_indicador
    FROM incidentes_sugeridos_ia s
    JOIN fuentes_noticias f ON f.id = s.fuente_id
    LEFT JOIN tipos_infraccion ti ON ti.id = s.tipo_infraccion_id
    WHERE s.estado = 'pendiente'
    ORDER BY s.creado_en DESC
  `)
  res.json(rows)
})

sugerenciasRouter.post('/:id/rechazar', async (req, res) => {
  await pool.query(`UPDATE incidentes_sugeridos_ia SET estado = 'rechazado' WHERE id = $1`, [req.params.id])
  res.status(204).end()
})

// Promueve la sugerencia a un incidente real (mismo flujo que "Registrar
// Incidente" desde una alerta de monitoreo). Si la IA no determinó un tipo
// de infracción, el analista debe indicarlo en el body.
sugerenciasRouter.post('/:id/aprobar', async (req, res) => {
  const { id } = req.params
  const tipoInfraccionIdOverride: string | undefined = req.body?.tipo_infraccion_id

  const { rows } = await pool.query(
    `SELECT s.*, ti.gravedad_sugerida
     FROM incidentes_sugeridos_ia s
     LEFT JOIN tipos_infraccion ti ON ti.id = s.tipo_infraccion_id
     WHERE s.id = $1 AND s.estado = 'pendiente'`,
    [id],
  )
  const sugerencia = rows[0]
  if (!sugerencia) {
    res.status(404).json({ error: 'Sugerencia no encontrada o ya procesada' })
    return
  }

  const tipoInfraccionId = tipoInfraccionIdOverride ?? sugerencia.tipo_infraccion_id
  if (!tipoInfraccionId) {
    res.status(400).json({ error: 'Debe indicarse tipo_infraccion_id: la IA no pudo determinarlo' })
    return
  }

  const { rows: incidenteRows } = await pool.query(
    `INSERT INTO incidentes (tipo_infraccion_id, descripcion, gravedad, estado, fecha_deteccion)
     VALUES ($1, $2, $3, 'sospechoso', now())
     RETURNING id, codigo`,
    [tipoInfraccionId, sugerencia.resumen ?? sugerencia.titular, sugerencia.gravedad_sugerida ?? 'medio'],
  )
  const incidente = incidenteRows[0]

  await pool.query(`UPDATE incidentes_sugeridos_ia SET estado = 'aprobado', incidente_id = $1 WHERE id = $2`, [
    incidente.id,
    id,
  ])

  res.json({ incidente })
})

File: server/app/routes/sugerencias.py
from fastapi import APIRouter, Body, HTTPException
from starlette import status

from ..db import ejecutar_sp, get_cursor

router = APIRouter()


# Cola de Validación IA: sugerencias pendientes de revisión por un analista.
@router.get("")
def listar():
    with get_cursor() as cur:
        ejecutar_sp(cur, "SELECT * FROM sp_sugerencias_listar_pendientes()")
        return cur.fetchall()


@router.post("/{sugerencia_id}/rechazar", status_code=status.HTTP_204_NO_CONTENT)
def rechazar(sugerencia_id: str):
    with get_cursor() as cur:
        ejecutar_sp(cur, "SELECT sp_sugerencias_rechazar(%s)", (sugerencia_id,))


# Promueve la sugerencia a un incidente real (mismo flujo que "Registrar
# Incidente" desde una alerta de monitoreo). Si la IA no determinó un tipo
# de infracción, el analista debe indicarlo en el body.
@router.post("/{sugerencia_id}/aprobar")
def aprobar(sugerencia_id: str, body: dict = Body(default={})):
    tipo_infraccion_id_override = body.get("tipo_infraccion_id")

    with get_cursor() as cur:
        ejecutar_sp(cur, "SELECT * FROM sp_sugerencias_obtener_pendiente(%s)", (sugerencia_id,))
        sugerencia = cur.fetchone()

    if not sugerencia:
        raise HTTPException(status_code=404, detail={"error": "Sugerencia no encontrada o ya procesada"})

    tipo_infraccion_id = tipo_infraccion_id_override or sugerencia.get("tipo_infraccion_id")
    if not tipo_infraccion_id:
        raise HTTPException(status_code=400, detail={"error": "Debe indicarse tipo_infraccion_id: la IA no pudo determinarlo"})

    with get_cursor() as cur:
        ejecutar_sp(
            cur,
            "SELECT * FROM sp_incidentes_registrar_desde_sugerencia(%s, %s, %s)",
            (
                tipo_infraccion_id,
                sugerencia.get("resumen") or sugerencia.get("titular"),
                sugerencia.get("gravedad_sugerida") or "medio",
            ),
        )
        incidente = cur.fetchone()

    with get_cursor() as cur:
        ejecutar_sp(cur, "SELECT sp_sugerencias_aprobar_vincular(%s, %s)", (sugerencia_id, incidente["id"]))

    return {"incidente": incidente}
